Git-commit: 12ba58af46e5973822f1cfaaf5f29a2a17a523bf
From: Joe Thornber <ejt@redhat.com>
Date: Fri, 27 Jul 2012 15:08:15 +0100
Subject: [PATCH] dm thin metadata: introduce dm_pool_metadata_set_read_only
Patch-mainline: v3.6-rc1
Reference: FATE#313903

Introduce dm_pool_metadata_set_read_only to put the underlying block
manager into read-only mode.

Signed-off-by: Joe Thornber <ejt@redhat.com>
Signed-off-by: Mike Snitzer <snitzer@redhat.com>
Signed-off-by: Alasdair G Kergon <agk@redhat.com>
Acked-by: NeilBrown <nfbrown@suse.com>

---
 drivers/md/dm-thin-metadata.c |   20 ++++++++++++++++----
 drivers/md/dm-thin-metadata.h |    6 ++++++
 2 files changed, 22 insertions(+), 4 deletions(-)

--- linux-3.0-SLE11-SP3.orig/drivers/md/dm-thin-metadata.c
+++ linux-3.0-SLE11-SP3/drivers/md/dm-thin-metadata.c
@@ -184,6 +184,7 @@ struct dm_pool_metadata {
 	uint64_t trans_id;
 	unsigned long flags;
 	sector_t data_block_size;
+	bool read_only:1;
 };
 
 struct dm_thin_device {
@@ -807,6 +808,7 @@ struct dm_pool_metadata *dm_pool_metadat
 	init_rwsem(&pmd->root_lock);
 	pmd->time = 0;
 	INIT_LIST_HEAD(&pmd->thin_devices);
+	pmd->read_only = false;
 	pmd->bdev = bdev;
 	pmd->data_block_size = data_block_size;
 
@@ -849,10 +851,12 @@ int dm_pool_metadata_close(struct dm_poo
 		return -EBUSY;
 	}
 
-	r = __commit_transaction(pmd);
-	if (r < 0)
-		DMWARN("%s: __commit_transaction() failed, error = %d",
-		       __func__, r);
+	if (!pmd->read_only) {
+		r = __commit_transaction(pmd);
+		if (r < 0)
+			DMWARN("%s: __commit_transaction() failed, error = %d",
+			       __func__, r);
+	}
 
 	__destroy_persistent_data_objects(pmd);
 	kfree(pmd);
@@ -1587,3 +1591,11 @@ int dm_pool_resize_data_dev(struct dm_po
 
 	return r;
 }
+
+void dm_pool_metadata_read_only(struct dm_pool_metadata *pmd)
+{
+	down_write(&pmd->root_lock);
+	pmd->read_only = true;
+	dm_bm_set_read_only(pmd->bm);
+	up_write(&pmd->root_lock);
+}
--- linux-3.0-SLE11-SP3.orig/drivers/md/dm-thin-metadata.h
+++ linux-3.0-SLE11-SP3/drivers/md/dm-thin-metadata.h
@@ -174,6 +174,12 @@ int dm_pool_get_data_dev_size(struct dm_
  */
 int dm_pool_resize_data_dev(struct dm_pool_metadata *pmd, dm_block_t new_size);
 
+/*
+ * Flicks the underlying block manager into read only mode, so you know
+ * that nothing is changing.
+ */
+void dm_pool_metadata_read_only(struct dm_pool_metadata *pmd);
+
 /*----------------------------------------------------------------*/
 
 #endif
